Bugfix permission requesting

Tobis93 opened the pull request:
Ticket #21

Tobis93 committed:
Bugfix for Permission-Requesting:
- App asked for both permissions (Location and ExternalStorage) at the same time, but is only allowed to ask for one at a time
Issues:
- Currently, the app asks the user again and again for permissions --> Necessary?
- There is a "Bug" in Android, where when a User chooses "Don't ask again" for a permission, the app can't ever ask again for it. If the user chooses this, the app will close and show a toast that says he needs to grant the permission in the preferences. (Needs to be discussed)

Tobis93 committed:
Added Dialog for Missing Permission with the "Don't Ask Again"-Flag that offers the User to bring him directly to the App-Preferences
If the user cancels the dialog, all buttons will be hidden

Tobis93 committed:
- Enhanced Dialog with shorter Texts
- If the user chooses "Cancel", the app will close (because we can't deliver functionality without the permissions)
- Added texts from the dialog to the strings.xml

Tobis93 committed:
- Removed unnecessary hiding of the buttons before the app closes after the user choose "Cancel"

anykeyjunkie commented:
permissions need to be checked at the start of every activity. maybe we should setup an application class so we have one central starting point?

Tobis93 committed:
- added new Activity for RequestPermission-Method-Handling from which all other Activities inherit the methods
- Strange Bug that app can't request the correct status of the "Should I ask for permissions?"-Method on start of the app appeared. Therefore, the app currently directly shows the Dialog

Tobis93 committed:
-"Fixed" the Bug that the app would show Dialog directly by shifting the Test to other locations

Tobis93 committed:
Merge branch 'master_lab' into bugfix_permission_requesting

Tobis93 commented:
Fix should now work as wished.
Currently, the App-Close on "Cancel" in the Permission-Dialog is some kind of stuttering, maybe needs some adjustment we can discuss in our next session

anykeyjunkie commented on app/src/main/java/seemoo/fitbit/activities/MainActivity.java:
```diff
@@ -34,13 +38,14 @@
     private static final int REQUEST_ACCESS_FINE_LOCATION = 1;
-    private static final int REQUEST_EXTERNAL_STORAGE = 1;
+    private static final int REQUEST_EXTERNAL_STORAGE = 2;
+    private static final int REQUEST_APP_SETTINGS = 1;
```
both could just use a protected variant from RequestPermissionsActivity, or am I wrong?

Tobis93 replied:
Should work, going to test and adjust today.

Tobis93 replied:
Not necessary anymore because of other changes

anykeyjunkie commented on app/src/main/java/seemoo/fitbit/activities/MainActivity.java:
```diff
@@ -270,4 +243,64 @@ private void scanForLastDevice(){
+     * Checks if the user granted permission to access fine location.
+     */
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
```
can we not come up with a generic variant in RequestPermissionsActivity, so we do not have to copy this into every activity?

Tobis93 replied:
If I remember correctly, putting it in RequestPermissionsActivity doesn't work (for whatever reason). 
I'm going to test it again during our session today.

Tobis93 replied:
Did work, moved it into the RequestPermissionsActivity

anykeyjunkie commented on app/src/main/java/seemoo/fitbit/activities/WorkActivity.java:
```diff
@@ -366,6 +380,18 @@ public void onItemClick(AdapterView<?> parent, View view, int position, long id)
+    @Override
+    protected void onRestart() {
+        super.onRestart();
+        requestPermissionsLocation();
```
could we just put requestPermissionsLocation in onRestart of RequestPermissionsActivity ?

Tobis93 replied:
Did work, moved it into the requestPermissionsActivity

Tobis93 committed:
Adjusted Permission-Requesting corresponding to Johannes' Feedback:
- onRequestPermissionsResult implemented only in RequestPermissionsActivity
- Moved requesting for permissions on Restart of the App into the onRestart-Method of the RequestPermissionsActivity

anykeyjunkie committed:
if "cancel" is pressed in "but we need permissions!" dialog, activity is now finished. reboot loop earlier

anykeyjunkie commented:
@Tobis93 I had a problem when I tapped "Close" in the "Please give the permissions" warning dialog: app was constantly restarting, so no proper close. I made a commit to fix this. this solves as well the dependency on the intent parameter that we would need to check in every activity start otherwise. If you agree, please merge :)

Tobis93 commented:
@anykeyjunkie I thought I fixed the restart issue... 
However, the finish()-solution only works in the first activity, if you are in the second (third etc.) activity on the stack, it only closes the current activity.
Going to look for a fix asap.

anykeyjunkie commented:
Yes you are right, finish() closes only the top activity. This was OK in my eyes, though, since the error dialog pops up again and finally the app closes in the end. I could retest as well with your version, but on my phone after tapping "Close/Cancel" on the warning dialog, the app was in a restarting loop.

Tobis93 commented:
I'm fine with agreeing on the finish()-method.
I just retested "my" method. Seems to be either a vendor or an android-version thing:
It worked as intended on my Huawei (closing needed some time, but well), but it didn't work on the Motorola of my girlfriend. There it wasn't even a loop, it was just the blank app without anything showing. 

So, to sum up, do we want to agree on the finish()-method?

anykeyjunkie commented:
yes blank app was shown to me as well, I saw in the logcat that the process number was increasing all the time so I thought it might be restarting.
yes, I'm fine with finish() :)
